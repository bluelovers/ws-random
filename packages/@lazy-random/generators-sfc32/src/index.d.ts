import { RNGCore } from '@lazy-random/rng-abstract-core';
import { ISeedInputFromStringOrNumberOrArray } from '@lazy-random/seed-algorithm';
export declare type IRNGSfc32SeedTypes = readonly [number, number, number, number];
export declare class RNGSfc32 extends RNGCore {
    protected readonly _rng: () => number;
    protected readonly _seed: IRNGSfc32SeedTypes;
    constructor(seed?: ISeedInputFromStringOrNumberOrArray, opts?: any, ...argv: any);
    protected _init(seed: ISeedInputFromStringOrNumberOrArray, opts?: any, ...argv: any): void;
    seed(seed?: ISeedInputFromStringOrNumberOrArray, opts?: any, ...argv: any): void;
    get seedable(): boolean;
    next(): number;
    get name(): string;
}
export default RNGSfc32;
