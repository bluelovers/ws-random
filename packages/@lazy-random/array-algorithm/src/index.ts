
export * from './swapAlgorithm';
export * from './array_rebase';
