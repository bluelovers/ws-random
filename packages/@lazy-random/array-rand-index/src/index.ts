import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
