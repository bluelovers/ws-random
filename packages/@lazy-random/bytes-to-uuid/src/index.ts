import { BYTE_TO_HEX_TO_LOWER_CASE } from '@lazy-random/shared-lib';

export function _createBytesToUuidFn(bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	return (buf: ArrayLike<number>, offset?: number) =>
	{
		let i = offset || 0;
		// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
		return ([
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
		]).join('');
	}
}

/**
 * @see https://github.com/kelektiv/node-uuid/blob/master/lib/bytesToUuid.js
 */
export function bytesToUuid(buf: ArrayLike<number>, offset?: number, bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	let i = offset || 0;
	// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
	return ([
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
	]).join('');
}

export default bytesToUuid
