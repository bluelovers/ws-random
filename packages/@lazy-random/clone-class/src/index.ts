/**
 * @todo support typescript
 */
export function getClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o;

	if (thisArgv instanceof RNGClass)
	{
		// @ts-ignore
		o = (thisArgv.__proto__.constructor)
	}
	else
	{
		o = RNGClass
	}

	return o
}

/**
 * @todo support typescript
 */
export function cloneClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o = getClass(RNGClass, thisArgv, ...argv)

	// @ts-ignore
	return new o(...argv)
}

export default cloneClass
