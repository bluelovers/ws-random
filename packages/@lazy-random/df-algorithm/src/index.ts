
export { dfBates } from './bates'

export { dfBernoulli } from './bernoulli'
export { dfBinomial } from './binomial'
export { dfExponential } from './exponential'
export { dfGeometric } from './geometric'

export { dfIrwinHall } from './irwin-hall'
export { dfLogNormal } from './log-normal'

export { dfNormal } from './normal'
export { dfPareto } from './pareto'
