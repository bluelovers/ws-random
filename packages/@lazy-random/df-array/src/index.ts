
export { dfArrayIndex } from './array-index'
export { dfArrayShuffle } from './array-shuffle'

export { dfArrayUnique } from './array-unique'
export { dfArrayFill } from './array-fill'

