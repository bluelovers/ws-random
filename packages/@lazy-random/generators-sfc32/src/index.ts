import { df_sfc32 } from '@lazy-num/float-algorithm';
import { _MathRandom } from '@lazy-random/original-math-random';
import { df_xfnv1a, doubleToIEEE } from '@lazy-random/seed-algorithm';
import { RNGCore } from '@lazy-random/rng-abstract-core';
import { ITSValueOrArrayMaybeReadonly } from 'ts-type/lib/type/base';

export type IRNGSfc32AllowedSeedInputTypes = ITSValueOrArrayMaybeReadonly<string | number>;

export type IRNGSfc32SeedTypes = readonly [number, number, number, number];

export function _handleSeed(seedInput: IRNGSfc32AllowedSeedInputTypes): IRNGSfc32SeedTypes
{
	let exists_zero = false;
	let s: [number, number];

	const seed = [seedInput ?? []].flat().slice(0, 4);

	for (let i = 0; i < 4; i++)
	{
		const type = typeof seed[i];

		if (type === 'string')
		{
			seed[i] = df_xfnv1a(`${seed[i]}#sfc32#${i}`)();
		}
		else if (type !== 'number')
		{
			exists_zero = true;
			seed[i] = void 0;
		}
		else
		{
			seed[i] = Math.abs(seed[i] as number);
		}

		if (!seed[i])
		{
			if (seed[i] === 0 && !exists_zero)
			{
				exists_zero = true;
			}
			else
			{
				s ??= doubleToIEEE(_MathRandom());
				// @ts-ignore
				seed[i] = s.pop();
				if (!s.length)
				{
					s = void 0;
				}
			}
		}
	}

	return seed as any;
}

export class RNGSfc32 extends RNGCore
{
	protected readonly _rng: () => number;
	protected readonly _seed: IRNGSfc32SeedTypes;

	constructor(seed?: IRNGSfc32AllowedSeedInputTypes, opts?: any, ...argv: any)
	{
		super(seed, opts, ...argv);
		this._init(seed, opts, ...argv);
	}

	protected override _init(seed: IRNGSfc32AllowedSeedInputTypes, opts?: any, ...argv: any)
	{
		seed = _handleSeed(seed);

		// @ts-ignore
		this._seed = seed;
		// @ts-ignore
		this._rng = df_sfc32(...seed);
	}

	public override seed(seed?: IRNGSfc32AllowedSeedInputTypes, opts?: any, ...argv: any)
	{
		return this._init(seed, opts, ...argv);
	}

	public override get seedable(): boolean
	{
		return true;
	}

	next()
	{
		return this._rng();
	}

	override get name()
	{
		return 'sfc32';
	}
}

export default RNGSfc32;
