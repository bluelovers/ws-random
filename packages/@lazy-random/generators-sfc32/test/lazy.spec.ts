import { RNGSfc32 } from '../src/index';

test(`[0, 2654435769, 1013904242, 3668340011]`, () =>
{
	let rng = new RNGSfc32([0, 2654435769, 1013904242, 3668340011]);

	let actual: number[] = [];

	for (let i = 0; i < 16; i++)
	{
		actual.push(rng.next())
	}

	expect((rng as any)._seed).toHaveLength(4);
	expect(rng).toMatchSnapshot();
	expect(actual).toMatchSnapshot();

});

test(`['apple','apple','apple','apple','apple']`, () =>
{
	let rng = new RNGSfc32(['apple', 'apple', 'apple', 'apple', 'apple']);

	let actual: number[] = [];

	for (let i = 0; i < 16; i++)
	{
		actual.push(rng.next())
	}

	expect((rng as any)._seed).toHaveLength(4);
	expect(rng).toMatchSnapshot();
	expect(actual).toMatchSnapshot();

});
