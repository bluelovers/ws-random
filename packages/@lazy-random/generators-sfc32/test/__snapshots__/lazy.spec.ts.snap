// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`['apple','apple','apple','apple','apple'] 1`] = `
RNGSfc32 {
  "_rng": [Function],
  "_seed": Array [
    3467603136,
    171790564,
    2570879769,
    3571716107,
  ],
}
`;

exports[`['apple','apple','apple','apple','apple'] 2`] = `
Array [
  0.6789673378225416,
  0.25874294177629054,
  0.8167836032807827,
  0.6617898617405444,
  0.47150427498854697,
  0.8755843315739185,
  0.1887578519526869,
  0.8970381338149309,
  0.46180717274546623,
  0.6011948748491704,
  0.08018238912336528,
  0.2771920363884419,
  0.7973732207901776,
  0.028986769495531917,
  0.21976999030448496,
  0.920350780012086,
]
`;

exports[`[0, 2654435769, 1013904242, 3668340011] 1`] = `
RNGSfc32 {
  "_rng": [Function],
  "_seed": Array [
    0,
    2654435769,
    1013904242,
    3668340011,
  ],
}
`;

exports[`[0, 2654435769, 1013904242, 3668340011] 2`] = `
Array [
  0.4721359545364976,
  0.5977381586562842,
  0.10499598179012537,
  0.7162471688352525,
  0.7737048789858818,
  0.10252661304548383,
  0.5388140785507858,
  0.03130675223655999,
  0.28249518550001085,
  0.9021577208768576,
  0.7259290926158428,
  0.715998571831733,
  0.4378080458845943,
  0.681514490628615,
  0.9879769335966557,
  0.22878749459050596,
]
`;
